openmc/lattice.py

```diff
@@ -753,6 +764,336 @@ def is_valid_index(self, idx):
+                    pattern[1, 0] = getattr(self.universes[j][i-1], attribute)
+                    if j < self.shape[1] - 1:
+                        pattern[2, 0] = getattr(self.universes[j+1][i-1],
+                                                attribute)
+
+                if j == 0:
+                    if len(lattice_neighbors) > 0:
+                        pattern[0, 1] = lattice_neighbors[1]
+                        if i != 0:
+                            pattern[0, 0] = lattice_neighbors[1]
+                        if i != self.shape[0] - 1:
+                            pattern[0, 2] = lattice_neighbors[1]
+                    else:
+                        pattern[0, :] = dummy_neighbor
+                else:
+                    if i > 0:
+                        pattern[0, 0] = getattr(self.universes[j-1][i-1],
+                                                attribute)
+                    pattern[0, 1] = getattr(self.universes[j-1][i], attribute)
+                    if i < self.shape[0] - 1:
+                        pattern[0, 2] = getattr(self.universes[j-1][i+1],
+                                                attribute)
+
+                if i == self.shape[0] - 1:
+                    if len(lattice_neighbors) > 0:
+                        pattern[:, 2] = lattice_neighbors[4]
+                        if j == 0:
+                            pattern[0, 2] = lattice_neighbors[2]
+                        elif j == self.shape[1] - 1:
+                            pattern[2, 2] = lattice_neighbors[7]
+                    else:
+                        pattern[:, 2] = dummy_neighbor
+                else:
+                    if j > 0:
+                        pattern[0, 2] = getattr(self.universes[j-1][i+1],
+                                                attribute)
+                    pattern[1, 2] = getattr(self.universes[j][i+1], attribute)
+                    if j < self.shape[1] - 1:
+                        pattern[2, 2] = getattr(self.universes[j+1][i+1],
+                                                attribute)
+
+                if j == self.shape[1] - 1:
+                    if len(lattice_neighbors) > 0:
+                        pattern[2, 1] = lattice_neighbors[6]
+                        if i != 0:
+                            pattern[2, 0] = lattice_neighbors[6]
+                        if i != self.shape[0] - 1:
+                            pattern[2, 2] = lattice_neighbors[6]
+                    else:
+                        pattern[2, :] = dummy_neighbor
+                else:
+                    if i > 0:
+                        pattern[2, 0] = getattr(self.universes[j+1][i-1],
+                                                attribute)
+                    pattern[2, 1] = getattr(self.universes[j+1][i], attribute)
+                    if i < self.shape[0] - 1:
+                        pattern[2, 2] = getattr(self.universes[j+1][i+1],
+                                                attribute)
```

Comment:
All of these nested `if`s to handle the special cases make the code kind of tough to read. And it will get worse if anyone extends this to 3D. I think something you can do to simplify the code would be to write two locally defined functions (define them inside this function itself) that take `i` and `j` values and return the appropriate universe. One function would handle the case for `len(lattice_neighbors) > 0` and another would handle the other case. Then you can just check `len(lattice-neighbors)` once and alias the appropriate function for use in the `for` loops.

Reply:
Agreed! It's much nicer that way. 
I didnt exactly create a separate function of the case that neighbors arent specified, I made:
 - one function for neighbors outside the lattice, which got the len(lattice_neighbors) if statement out
- one function for neighbors inside the lattice